Modules/main.c

```diff
@@ -228,6 +228,152 @@ pymain_import_readline(const PyConfig *config)
+    Py_ssize_t candidate_len = 0;
+
+    for (char *iter = start; iter < end; ++iter) {
+        char *line_start = iter;
+        char *leading_whitespace_end = NULL;
+
+        // scan the whole line
+        while (iter < end && *iter != '\n') {
+            if (!leading_whitespace_end && *iter != ' ' && *iter != '\t') {
+                if (iter == line_start) {
+                    // some line has no indent, fast exit!
+                    return bytes;
+                }
+                leading_whitespace_end = iter;
+            }
+            ++iter;
+        }
+
+        // if this line has all white space, skip it
+        if (!leading_whitespace_end) {
+            continue;
+        }
+
+        if (!candidate_start) {
+            candidate_start = line_start;
+            candidate_len = leading_whitespace_end - line_start;
+            assert(candidate_len > 0);
+        } else {
+            /* We then compare with the current longest leading whitespace.
+
+               [line_start, leading_whitespace_end) is the leading whitespace of
+               this line,
+
+               [candidate_start, candidate_start + candidate_len)
+               is the leading whitespace of the current longest leading
+               whitespace. */
+            Py_ssize_t new_candidate_len = 0;
+
+            for (char *candidate_iter = candidate_start,
+                      *line_iter = line_start;
+                 candidate_iter < candidate_start + candidate_len &&
+                 line_iter < leading_whitespace_end;
+                 ++candidate_iter, ++line_iter) {
+                if (*candidate_iter != *line_iter) {
+                    break;
+                }
+                ++new_candidate_len;
+            }
+
+            candidate_len = new_candidate_len;
+            if (candidate_len == 0) {
+                return bytes;
+            }
+        }
+    }
+
+    if (candidate_len == 0) {
+        return bytes;
+    }
+    assert(candidate_len > 0);
+
+    // trigger a dedent
+    char *p;
+    PyObject *new_bytes;
+    char *line_start;
+    Py_ssize_t new_line_len;
+    bool in_leading_space;
+    _PyBytesWriter writer;
+
+    _PyBytesWriter_Init(&writer);
+    p = _PyBytesWriter_Alloc(&writer, nchars);
+    if (p == NULL) {
+        goto error;
+    }
+
+    for (char *iter = start; iter < end; ++iter) {
+        line_start = iter;
+        in_leading_space = true;
+
+        // iterate over a line
+        while (iter < end && *iter != '\n') {
+            if (in_leading_space && *iter != ' ' && *iter != '\t') {
+                in_leading_space = false;
+            }
+            ++iter;
+        }
+
+        // invariant: *iter == '\n' or iter == end
+
+        // if this line has all white space, write '\n'
+        if (in_leading_space) {
+            p = _PyBytesWriter_Prepare(&writer, p, 1);
+            if (p == NULL) {
+                goto error;
+            }
+            *p++ = '\n';
```

Comment:
```suggestion
            *p++ = '\n';
```

since you allocated maximum required memory first, you don't need to call `_PyBytesWriter_Prepare` at all.